### PR TITLE
adds terms-of-use fields and license to json printer/parser

### PR DESCRIPTION
**Pull request by bmckinney**
# RFI Checklist

### 1. Related Issues

- #2794 + license and terms of access information not appearing in json
- #1991 Verify that API sets license on create dataset

---
### 2. Pull Request Checklist

- [ ]  Functionality completed as described in FRD
- [ ]  Dependencies, risks, assumptions in FRD addressed
- [ ]  Unit tests completed
- [ ]  Deployment requirements identified (e.g., SQL scripts, indexing)
- [ ]  Documentation completed
- [ ]  All code checkins completed

---
### 3. Review Checklist

_**After** the pull request has been submitted, fill out this section._

- [ ]  Code review completed or waived
- [ ]  Testing requirements completed
- [ ]  Usability testing completed or waived
- [ ]  Support testing completed or waived
- [ ]  Merged with develop branch and resolved conflicts

### CODE REVIEW

File: src/main/java/edu/harvard/iq/dataverse/util/json/JsonPrinter.java [hunk 1]
```diff
@@ -1,22 +1,8 @@
...
-import edu.harvard.iq.dataverse.DatasetDistributor;
-import edu.harvard.iq.dataverse.DatasetFieldType;
-import edu.harvard.iq.dataverse.DatasetField;
-import edu.harvard.iq.dataverse.DatasetFieldCompoundValue;
-import edu.harvard.iq.dataverse.DatasetFieldValue;
-import edu.harvard.iq.dataverse.DatasetVersion;
-import edu.harvard.iq.dataverse.Dataverse;
-import edu.harvard.iq.dataverse.DataverseContact;
-import edu.harvard.iq.dataverse.DataverseFacet;
+import edu.harvard.iq.dataverse.*;
```
Comment (pdurbin) [inline src/main/java/edu/harvard/iq/dataverse/util/json/JsonPrinter.java]: No wildcard (`*`) imports, please.

File: src/main/java/edu/harvard/iq/dataverse/util/json/JsonPrinter.java [hunk 2]
```diff
@@ -503,4 +515,10 @@ public static JsonObjectBuilder json( AuthenticationProviderRow aRow ) {
         }
         return bld;
     }
+
+    public static void addNullAware(JsonObjectBuilder builder, String name, String value)  {
```
Comment (pdurbin) [inline src/main/java/edu/harvard/iq/dataverse/util/json/JsonPrinter.java]: Is this method necessary? Can we make use of @michbarsinai 's https://github.com/IQSS/dataverse/blob/v4.3/src/main/java/edu/harvard/iq/dataverse/util/json/NullSafeJsonBuilder.java instead?

**Comment by pdurbin** [comment]
@bmckinney does this pull request allow me to **set** terms of use when creating a dataset? If not, could it?

@sekmiller you're actively looking at all this JSON reading/writing code by @michbarsinai for #2608 so you should probably weigh in on this pull request.

File: src/main/java/edu/harvard/iq/dataverse/util/json/JsonPrinter.java (hunk 2 above)
Comment (bmckinney) [inline src/main/java/edu/harvard/iq/dataverse/util/json/JsonPrinter.java]: Sorry, I didn't notice that utility. I'll revise accordingly.

**Comment by sekmiller** [comment]
These changes have been incorporated into the Harvesting/Export Branch

**Comment by pdurbin** [comment]
@sekmiller good because this branch has merge conflicts anyway. :smile: 

However! Last I knew @bmckinney is still using the branch this pull request is based on (2794-add-terms-of-use-license-to-json) when he builds a war file so let's touch base with him to see if this pull request can be closed and that branch deleted.

File: src/main/java/edu/harvard/iq/dataverse/util/json/JsonParser.java [hunk 3]
```diff
@@ -211,9 +229,17 @@ public DatasetVersion parseDatasetVersion(JsonObject obj, DatasetVersion dsv) th
             terms.setContactForAccess(obj.getString("contactForAccess", null));
             terms.setSizeOfCollection(obj.getString("sizeOfCollection", null));
             terms.setStudyCompletion(obj.getString("studyCompletion", null));
-            /* License???*/
+
+            // License: only CC0 or NONE is currently supported
+            TermsOfUseAndAccess.License license = TermsOfUseAndAccess.License.NONE;
```
Comment (pdurbin) [inline src/main/java/edu/harvard/iq/dataverse/util/json/JsonParser.java]: @sekmiller and I talked about this the other day and we think by default the license should be `CC0` to match what the GUI does. (Yes, this is a change in behavior, which is what #1991 is about... how through the native API you end up with a license of "null" which is never what we want.) Best would probably be to reference the default license via the existing static variable like this:

    TermsOfUseAndAccess.License license = TermsOfUseAndAccess.defaultLicense;

Obviously, the logic below would need to change to change `CC0` to `NONE` when you see fit (when you supply "NONE" in the JSON).

**Comment by pdurbin** [comment]
I just put this in the "Code Review" column in Waffle because I just left a comment in the code to capture a discussion @sekmiller and I had about this pull request the other day: https://github.com/IQSS/dataverse/pull/3082/files#r72855138

@bmckinney and @scolapasta please check it out.

**Comment by scolapasta** [comment]
FYI, I am planning to confirm with @bmckinney that we can close this PR since this code has been added to the export branch (just want to confirm with him to make sure we're not missing something).

Also, will ask @mcrosas about CC0 or none as default. I'm not as convinced that it should be CC0 because when going through the UI, you *see* that it is being set as default. But again will see what @mcrosas thinks.

**Comment by pdurbin** [comment]
@scolapasta that's fine but please be advised that @sekmiller was planning on implementing the change we talked about (having `CC0` rather than `null` persisted in the database when nothing is supplied in the JSON, a fix for #1991) in the harvesting/export branch, I believe. I'm not sure if he did this yet or not. He was going to ping you after he was done and before moving #1991 to QA.

**Comment by bmckinney** [comment]
@scolapasta & @pdurbin It's ok to close this PR.

**Comment by pdurbin** [comment]
@bmckinney heads up that behavior has changed in 4.5 regarding license and creating datasets via native API. Please see https://github.com/IQSS/dataverse/issues/1991#issuecomment-244437123

@scolapasta if you're going to chat with @mcrosas about licenses anyway, please note that I believe @Venki18 is interested having "CC BY" as the default license rather than "CC0". (@Venki18 please feel free to create a new GitHub issue about this and thanks for creating #3253 and friends. 😄 )

**Comment by mheppler** [comment]
FYI: There is already an issue tracking the "CC BY":  License: Multiple Options for Licensing #1753

**Comment by bmckinney** [comment]
before we close this, SBGrid needs to be able to get/set dataverseType via json:

in JsonParser:

        dv.setDataverseType(Dataverse.DataverseType.UNCATEGORIZED); // default
        if (jobj.containsKey("dataverseType")) {
            for (Dataverse.DataverseType dvtype : Dataverse.DataverseType.values()) {
                if (dvtype.name().equalsIgnoreCase(jobj.getString("dataverseType"))) {
                    dv.setDataverseType(dvtype);
                }
            }
        }

in JsonPrinter:

    public static JsonObjectBuilder json(Dataverse dv) {
        JsonObjectBuilder bld = jsonObjectBuilder()
                .add("id", dv.getId())
                .add("alias", dv.getAlias())
                .add("name", dv.getName())
                .add("affiliation", dv.getAffiliation())
                .add("dataverseContacts", json(dv.getDataverseContacts()))
                .add("permissionRoot", dv.isPermissionRoot())
                .add("description", dv.getDescription())
                .add("dataverseType", dv.getDataverseType().name());
 ...

**Comment by pdurbin** [comment]
@bmckinney thanks for the heads up. As you're well aware (since you wrote the bulk of the code), we opened pull request #3373 to handle dataverseType ("category" in the GUI) so I'm closing this pull request.

**Comment by pdurbin** [comment]
> FYI: There is already an issue tracking the "CC BY": License: Multiple Options for Licensing #1753

@mheppler thanks, I finally got around to moving my comment there: https://github.com/IQSS/dataverse/issues/1753#issuecomment-255585621

**Pull request closed without merging**